File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/EditNotesRecyclerViewAdapter.java
package org.scp.gymlog.ui.common.dialogs;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.databinding.ListElementFragmentTextBinding;

import java.util.List;
import java.util.function.Consumer;

public class EditNotesRecyclerViewAdapter extends RecyclerView.Adapter<EditNotesRecyclerViewAdapter.ViewHolder> {

    private final List<String> values;
    private final Consumer<String> onClick;

    public EditNotesRecyclerViewAdapter(List<String> values, Consumer<String> onClick) {
        this.values = values;
        this.onClick = onClick;
    }

    @Override
    public EditNotesRecyclerViewAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new EditNotesRecyclerViewAdapter.ViewHolder(
                ListElementFragmentTextBinding.inflate(
                        LayoutInflater.from(parent.getContext()), parent, false
                )
        );
    }

    @Override
    public void onBindViewHolder(final EditNotesRecyclerViewAdapter.ViewHolder holder, int position) {
        holder.mContentView.setText(values.get(position));
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final TextView mContentView;

        public ViewHolder(ListElementFragmentTextBinding binding) {
            super(binding.getRoot());
            mContentView = binding.content;

            itemView.setOnClickListener(view ->
                onClick.accept(mContentView.getText().toString())
            );
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mContentView.getText() + "'";
        }
    }
}

File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/EditNotesRecyclerViewAdapter.kt
package org.scp.gymlog.ui.common.dialogs

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.databinding.ListitemSelectableTextBinding

class EditNotesRecyclerViewAdapter(
    private val values: List<String>,
    preSelected: List<String> = emptyList(),
    private val onClick: (String, Boolean) -> Unit
) : RecyclerView.Adapter<EditNotesRecyclerViewAdapter.ViewHolder>() {

    private val initialSelects = preSelected.toMutableList()

    private val clearActions = mutableListOf<Runnable>()
    fun clearSelects() = clearActions.forEach { it.run() }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            ListitemSelectableTextBinding.inflate(
                LayoutInflater.from(parent.context), parent, false
            )
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val text = values[position]
        holder.mContentView.text = text
        holder.check = initialSelects.contains(text)
    }

    override fun getItemCount(): Int {
        return values.size
    }

    inner class ViewHolder(
        binding: ListitemSelectableTextBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        val mContentView: TextView = binding.content
        private val mCheck = binding.check

        var check: Boolean
            get() = mCheck.visibility == View.VISIBLE
            set(value) { mCheck.visibility = if(value) View.VISIBLE else View.INVISIBLE }

        init {
            clearActions.add {
                if (check) check = false
            }

            itemView.setOnClickListener {
                val value = !check
                val text = mContentView.text.toString()
                check = value
                onClick(text, value)
                if (!value) {
                    initialSelects.remove(text)
                }
            }
        }

        override fun toString(): String {
            return super.toString() + " '" + mContentView.text + "'"
        }
    }
}